File: code_snippets/java/src/main/java/develop/workflows/WorkflowSubmitter.java
package develop.workflows;

import dev.restate.sdk.client.Client;
import dev.restate.sdk.client.SendResponse;
import dev.restate.sdk.common.Output;

public class WorkflowSubmitter {

  public void submitWorkflow(Email email) {
    // <start_submit>
    Client restateClient = Client.connect("http://localhost:8080");
    SendResponse handle = SignupWorkflowClient.fromClient(restateClient, "someone").submit(email);
    // <end_submit>

    // <start_query>
    String status = SignupWorkflowClient.fromClient(restateClient, "someone").getStatus();
    // <end_query>

    // <start_interact>
    // Option 1: attach and wait for result
    boolean result =
        SignupWorkflowClient.fromClient(restateClient, "someone").workflowHandle().attach();

    // Option 2: peek to check if ready
    Output<Boolean> peekOutput =
        SignupWorkflowClient.fromClient(restateClient, "someone").workflowHandle().getOutput();
    if (peekOutput.isReady()) {
      boolean result2 = peekOutput.getValue();
    }
    // <end_interact>

  }
}

File: code_snippets/kotlin/src/main/kotlin/develop/workflows/WorkflowSubmitter.kt
package develop.workflows

import dev.restate.client.Client
import dev.restate.client.kotlin.*

class WorkflowSubmitter {

  suspend fun submitWorkflow(email: Email) {
    // <start_submit>
    val restate = Client.connect("http://localhost:8080")
    val sendResponse = SignupWorkflowClient.fromClient(restate, "someone").submit(email)
    // <end_submit>

    // <start_query>
    val status = SignupWorkflowClient.fromClient(restate, "someone").getStatus()
    // <end_query>

    // <start_interact>

    // <end_query>

    // <start_interact>
    // Option 1: attach and wait for result
    val result =
        SignupWorkflowClient.fromClient(restate, "someone")
            .workflowHandle()
            .attachSuspend()
            .response

    // Option 2: peek to check if ready
    val peekOutput =
        SignupWorkflowClient.fromClient(restate, "someone")
            .workflowHandle()
            .getOutputSuspend()
            .response
    if (peekOutput.isReady) {
      val result2 = peekOutput.value
    }
    // <end_interact>
  }
}
